src/client.ts

```diff
@@ -172,27 +172,31 @@ export class Client {
    *                               sort:
    *                                 field: "asc" | "desc"
    *                               sort: [{field: "asc|desc"}]
-   * @param  {Boolean}  wrap  Tell the client to return a class object instead of
-   *                             the raw value returned by the endpoint. Defaults to true
    *
    *
    * @return {Promise<Response>}
    */
-  public async search(
-    query: string | Query,
-    params?: DoofinderParameters,
-    wrap = true
-  ): Promise<Response | DoofinderResult> {
+  public async search(query: string | Query, params?: DoofinderParameters): Promise<Response | DoofinderResult> {
     const qs: string = this.buildSearchQueryString(query, params);
-    let response: Response;
-    if (query instanceof Query && query.items) {
-      response = await this.request(this.buildUrl('/getitems', qs), { items: query.items });
-    } else {
-      response = await this.request(this.buildUrl('/search', qs));
-    }
+    const response = await this.request(this.buildUrl('/search', qs));
+    const data = await response.json();
+    return new DoofinderResult(data);
+  }
+
+  /**
+   * Perform a get items request.
+   *
+   * @param {Query}   query   Query definition for get items request. This query
+   *                          must has the items list defined.
+   */
+  public async getItems(query: Query): Promise<Response | DoofinderResult> {
+    const qs: string = this.buildSearchQueryString(query);
```

Comment:
As I've said, this is too specific for search and I'm not sure it should be used here. 

In any case it needs a refactoring we can do later but the request done inside `getItems()` should not have any kind of `query` parameter and that method performs some checks for the query parameter.

Apart from that see this example with the `Query` object:

```js
const query = new Query();
query.searchItems([…]); // resets query
query.search('whatever'); // doesn't reset items
```

If you pass the previous `Query` object to `search()` nothing happens because the implementation does nothing with items. If you pass it to `getItems()` you may end with a `query` parameter which is not what the endpoint should accept.

Reply:
You can make a getItems with filters, that is why it's necessary to use buildSearchQueryString

Reply:
I know that, but what I try to say is that there's some validation there that should be inside `Query`. For me, the `buildSearchQueryString()` method should be something like this:

```typescript
// DoofinderParameters should include `query` and `hashid` if they're not there
public buildSearchQueryString(queryOrParams: Query|DoofinderParameters): string {
  let query: Query;

  if (queryOrParams instanceof Query) {
    query = queryOrParams;
  } else {
    query = new Query();
    query.setParameters(queryOrParams);
  }

  if (!query.hasParameter('hashid') && this.hashid) {
    query.setParameter('hashid', this.hashid);
  }

  return buildQueryString(query.getParams(true));
}
```

`Query.getParams(validate: boolean = false)` would have a `validate` parameter. If something is missing, like `hashid` it would throw a `QueryError`.

Internally `Query.getParams` would handle the types de-normalization (for me, `this.params.type` should always be an array):

```typescript
if (this.params.type.length < 2) {
  // undefined if length is 0, so OK because it should not appear in the qs
  params.type = this.params.type[0];
} else {
  // copy array
  params.type = this.params.type.slice();
}
```

I'm not sure if we have to set `query` to `' '` when we can just set it to `''`. We could force the `query_name` to `match_all` for text queries in that case unless a specific `query_name` is set in the query (this requires some thinking).

But in any case, normalizing should be done when setting parameters and de-normalizing should be done when dumping parameters.

I would rename `getParams` to `getParameters` or `dump` or `export` or something that says you are returning the whole query as an object that can serialize to a querystring.

Reply:
In fact, if there's no `hashid` in the `Query` instance and there's no `hashid` in the `Client` instance, the request would be wrong.

I'm not a fan of setting `hashid` in the `Client` at all and I would remove it from there (and that check would be removed too). For me, it's just a parameter, but there was a draw when we gave our opinions so we left it there.